analyzers/src/SonarAnalyzer.Common/Rules/CommentsShouldNotBeEmptyBase.cs

```diff
@@ -0,0 +1,126 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class CommentsShouldNotBeEmptyBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S4663";
+
+    protected abstract string GetCommentText(SyntaxTrivia trivia);
+    protected abstract bool IsValidTriviaType(SyntaxTrivia trivia);
+
+    protected abstract bool IsSimpleComment(SyntaxTrivia trivia);
+    protected abstract bool IsEndOfLine(SyntaxTrivia trivia);
+    protected abstract bool IsWhitespace(SyntaxTrivia trivia);
+
+    protected override string MessageFormat => "Remove this empty comment";
+
+    protected CommentsShouldNotBeEmptyBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterTreeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            foreach (var token in c.Tree.GetRoot().DescendantTokens())
+            {
+                CheckTrivia(c, token.LeadingTrivia);
+                CheckTrivia(c, token.TrailingTrivia);
```

Comment:
When Roslyn does the parsing it only ever fills LeadingTrivia if I'm not mistaken. The TrailingTrivia is used by CodeFixer to shuffle the trivia around when changing the SyntaxTree.
Given that it would be ok to check only leading trivia, but to be on the safe side we should still check both.

Reply:
I think most of the Trivia are indeed Leading, but for example here:
```csharp
//
class X { } // <-- trailing
```
You can find trailing trivia as well.

Reply:
Oh. Right. The last trivia needs to be trailing.